File: mobile_client/lib/pages/island_designs_page.dart
import 'package:flutter/material.dart';

import '../api/island_designs.dart';
import '../island/island_models.dart';
import '../main.dart' as app;
import '../theme.dart';
import '../widgets/glass.dart';

/// Settings tab for Dynamic Island designs (iOS only).
///
/// A single radio group picks what shows on the Dynamic Island: **Auto** (rules
/// decide), or a specific design pinned — including the built-in Voice + Coding
/// islands. Below, a management list toggles each entry's inclusion in Auto and
/// deletes custom designs. Designs themselves are authored by Jarvis via the
/// dynamic-island skill, not here.
class IslandDesignsPage extends StatefulWidget {
  const IslandDesignsPage({super.key});

  @override
  State<IslandDesignsPage> createState() => _IslandDesignsPageState();
}

class _IslandDesignsPageState extends State<IslandDesignsPage> {
  static const _autoKey = '__auto__';

  late final IslandApi _api = IslandApi(app.api);
  IslandCatalog _catalog = IslandCatalog.empty;
  bool _loading = true;
  String? _error;
  bool _busy = false;

  @override
  void initState() {
    super.initState();
    _load();
  }

  Future<void> _load() async {
    setState(() {
      _loading = true;
      _error = null;
    });
    try {
      final cat = await _api.fetchCatalog();
      if (!mounted) return;
      setState(() {
        _catalog = cat;
        _loading = false;
      });
    } catch (_) {
      if (!mounted) return;
      setState(() {
        _loading = false;
        _error = 'Could not load designs. Check the server connection.';
      });
    }
  }

  String get _selectedKey =>
      _catalog.selection.isAuto ? _autoKey : (_catalog.selection.pinnedId ?? _autoKey);

  Future<void> _guard(Future<void> Function() op) async {
    if (_busy) return;
    setState(() => _busy = true);
    try {
      await op();
      await _load();
    } catch (_) {
      if (mounted) {
        ScaffoldMessenger.of(context).showSnackBar(
          const SnackBar(content: Text('That didn’t go through. Try again.')),
        );
      }
    } finally {
      if (mounted) setState(() => _busy = false);
    }
  }

  Future<void> _select(String key) => _guard(() => key == _autoKey
      ? _api.setSelection('auto')
      : _api.setSelection('pinned', pinnedId: key));

  Future<void> _setEnabled(String id, bool v) =>
      _guard(() => _api.setRules(id, enabled: v));

  Future<void> _delete(IslandCatalogEntry e) async {
    final ok = await showDialog<bool>(
      context: context,
      builder: (_) => AlertDialog(
        title: Text('Delete “${e.name}”?'),
        content: const Text(
            'This removes the design. Jarvis can recreate it later via the '
            'dynamic-island skill.'),
        actions: [
          TextButton(
              onPressed: () => Navigator.pop(context, false),
              child: const Text('Cancel')),
          TextButton(
              onPressed: () => Navigator.pop(context, true),
              child: const Text('Delete')),
        ],
      ),
    );
    if (ok == true) await _guard(() => _api.deleteDesign(e.id));
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.transparent,
      extendBodyBehindAppBar: true,
      appBar: glassAppBar(context, title: 'Dynamic Island', back: true),
      body: AppBackground(
        child: SafeArea(
          child: _loading
              ? const Center(child: CircularProgressIndicator())
              : RefreshIndicator(
                  onRefresh: _load,
                  child: _buildBody(),
                ),
        ),
      ),
    );
  }

  Widget _buildBody() {
    final entries = _catalog.entries;
    final custom = entries.where((e) => !e.builtin).toList(growable: false);

    return ListView(
      padding: const EdgeInsets.fromLTRB(20, 8, 20, 32),
      children: [
        const _Intro(),
        const SizedBox(height: 22),
        if (_error != null) ...[
          _ErrorCard(_error!),
          const SizedBox(height: 18),
        ],

        // ── Selection (the radio group) ──
        glassSectionLabel('Show on Dynamic Island'),
        GlassGroup(children: [
          _RadioRow(
            value: _autoKey,
            groupValue: _selectedKey,
            icon: Icons.auto_mode_rounded,
            title: 'Auto',
            subtitle: 'Let rules pick the right island at the right time.',
            onChanged: _busy ? null : _select,
          ),
          for (var i = 0; i < entries.length; i++)
            _RadioRow(
              value: entries[i].id,
              groupValue: _selectedKey,
              icon: _iconFor(entries[i]),
              title: entries[i].name,
              subtitle: _entrySubtitle(entries[i]),
              onChanged: _busy ? null : _select,
              last: i == entries.length - 1,
            ),
        ]),
        const SizedBox(height: 26),

        // ── Manage (enable in Auto pool + delete) ──
        glassSectionLabel('In the Auto rotation'),
        GlassGroup(
          children: [
            for (var i = 0; i < entries.length; i++)
              _ManageRow(
                entry: entries[i],
                busy: _busy,
                onEnabled: (v) => _setEnabled(entries[i].id, v),
                onDelete: entries[i].builtin ? null : () => _delete(entries[i]),
                last: i == entries.length - 1,
              ),
          ],
        ),

        if (custom.isEmpty) ...[
          const SizedBox(height: 18),
          const _EmptyHint(),
        ],
      ],
    );
  }

  String _entrySubtitle(IslandCatalogEntry e) {
    final bits = <String>[];
    if (e.builtin) bits.add('Built-in');
    if (!e.enabled) bits.add('Off in Auto');
    bits.add('Priority ${e.priority}');
    return bits.join(' · ');
  }

  IconData _iconFor(IslandCatalogEntry e) {
    if (e.isVoice) return Icons.graphic_eq_rounded;
    if (e.isCoding) return Icons.code_rounded;
    return Icons.dashboard_customize_rounded;
  }
}

class _RadioRow extends StatelessWidget {
  const _RadioRow({
    required this.value,
    required this.groupValue,
    required this.icon,
    required this.title,
    required this.subtitle,
    required this.onChanged,
    this.last = false,
  });
  final String value, groupValue, title, subtitle;
  final IconData icon;
  final ValueChanged<String>? onChanged;
  final bool last;

  @override
  Widget build(BuildContext context) {
    final selected = value == groupValue;
    return Column(
      children: [
        InkWell(
          onTap: onChanged == null ? null : () => onChanged!(value),
          child: Padding(
            padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 12),
            child: Row(
              children: [
                Container(
                  width: 40,
                  height: 40,
                  decoration: BoxDecoration(
                    shape: BoxShape.circle,
                    color: JcTheme.glassFill,
                    border: Border.all(color: JcTheme.glassBorder),
                  ),
                  child: Icon(icon, size: 20, color: JcTheme.text),
                ),
                const SizedBox(width: 14),
                Expanded(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(title,
                          style: const TextStyle(
                              color: JcTheme.text,
                              fontSize: 15,
                              fontWeight: FontWeight.w600)),
                      const SizedBox(height: 2),
                      Text(subtitle,
                          style: const TextStyle(
                              color: JcTheme.muted, fontSize: 12)),
                    ],
                  ),
                ),
                Icon(
                  selected
                      ? Icons.radio_button_checked_rounded
                      : Icons.radio_button_unchecked_rounded,
                  color: selected ? JcTheme.primaryBlue : JcTheme.muted,
                ),
              ],
            ),
          ),
        ),
        if (!last)
          const Divider(height: 1, indent: 68, color: JcTheme.glassBorder),
      ],
    );
  }
}

class _ManageRow extends StatelessWidget {
  const _ManageRow({
    required this.entry,
    required this.busy,
    required this.onEnabled,
    required this.onDelete,
    this.last = false,
  });
  final IslandCatalogEntry entry;
  final bool busy;
  final ValueChanged<bool> onEnabled;
  final VoidCallback? onDelete;
  final bool last;

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        Padding(
          padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 6),
          child: Row(
            children: [
              Expanded(
                child: Text(entry.name,
                    style: const TextStyle(
                        color: JcTheme.text,
                        fontSize: 15,
                        fontWeight: FontWeight.w600)),
              ),
              if (onDelete != null)
                IconButton(
                  icon: const Icon(Icons.delete_outline_rounded,
                      color: JcTheme.muted),
                  onPressed: busy ? null : onDelete,
                  tooltip: 'Delete',
                ),
              Switch(
                value: entry.enabled,
                onChanged: busy ? null : onEnabled,
                activeThumbColor: Colors.white,
                activeTrackColor: JcTheme.primaryBlue,
              ),
            ],
          ),
        ),
        if (!last)
          const Divider(height: 1, indent: 14, color: JcTheme.glassBorder),
      ],
    );
  }
}

class _Intro extends StatelessWidget {
  const _Intro();
  @override
  Widget build(BuildContext context) => const Padding(
        padding: EdgeInsets.symmetric(horizontal: 4),
        child: Text(
          'Pick which Dynamic Island shows — or let Auto choose by time and '
          'context. Jarvis creates custom designs for you; they appear here.',
          style: TextStyle(color: JcTheme.muted, fontSize: 13, height: 1.4),
        ),
      );
}

class _EmptyHint extends StatelessWidget {
  const _EmptyHint();
  @override
  Widget build(BuildContext context) => Container(
        padding: const EdgeInsets.all(16),
        decoration: BoxDecoration(
          color: JcTheme.glassFill,
          borderRadius: BorderRadius.circular(16),
          border: Border.all(color: JcTheme.glassBorder),
        ),
        child: const Text(
          'No custom designs yet. Ask Jarvis to make one — e.g. “make a Dynamic '
          'Island that shows my next meeting”.',
          style: TextStyle(color: JcTheme.muted, fontSize: 13, height: 1.4),
        ),
      );
}

class _ErrorCard extends StatelessWidget {
  const _ErrorCard(this.message);
  final String message;
  @override
  Widget build(BuildContext context) => Container(
        padding: const EdgeInsets.all(14),
        decoration: BoxDecoration(
          color: Colors.red.withValues(alpha: 0.12),
          borderRadius: BorderRadius.circular(14),
          border: Border.all(color: Colors.red.withValues(alpha: 0.3)),
        ),
        child: Text(message,
            style: const TextStyle(color: JcTheme.text, fontSize: 13)),
      );
}

File: ios_app/JarvisCopilot/Copilot/More/IslandDesigns/IslandDesignsPage.swift
import SwiftUI

/// The "Dynamic Island" settings screen, ported from `pages/island_designs_page.dart`.
///
/// One radio group picks what shows: **Auto** (rules decide) or a specific entry
/// pinned, including the built-in Voice and Coding islands. Below it a management
/// list toggles each entry's place in the Auto rotation, nudges its priority and
/// deletes custom designs. Designs are authored by Jarvis via the dynamic-island
/// skill, never here.
struct IslandDesignsPage: View {
    @State private var store: IslandDesignsStore
    @State private var deleting: IslandCatalogEntry?

    /// See `SettingsPage.init` — a view's `init` isn't main-actor-isolated, so
    /// the store can't be a default argument.
    init(store: IslandDesignsStore? = nil) {
        _store = State(initialValue: store ?? MainActor.assumeIsolated {
            IslandDesignsStore.production()
        })
    }

    var body: some View {
        content
            .jcScreen("Dynamic Island")
            .task { if store.entries.isEmpty { store.load() } }
            .moreToast($store.toast)
            .alert("Delete this design?", isPresented: deletingBinding, presenting: deleting) { entry in
                Button("Cancel", role: .cancel) {}
                Button("Delete", role: .destructive) { Task { await store.delete(entry) } }
            } message: { entry in
                Text("This removes “\(entry.name)”. Jarvis can recreate it later via "
                   + "the dynamic-island skill.")
            }
    }

    private var deletingBinding: Binding<Bool> {
        Binding(get: { deleting != nil }, set: { if !$0 { deleting = nil } })
    }

    @ViewBuilder
    private var content: some View {
        if store.isLoading && store.entries.isEmpty {
            ProgressView().frame(maxWidth: .infinity, maxHeight: .infinity)
        } else {
            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    Text("Pick which Dynamic Island shows — or let Auto choose by time "
                       + "and context. Jarvis creates custom designs for you; they "
                       + "appear here.")
                        .font(.system(size: 13))
                        .foregroundStyle(JcTheme.muted)
                        .fixedSize(horizontal: false, vertical: true)
                        .padding(.horizontal, 4)
                        .padding(.bottom, 22)

                    if let message = store.errorMessage {
                        IslandErrorCard(message: message).padding(.bottom, 18)
                    }

                    GlassSectionLabel("Show on Dynamic Island")
                    GlassGroup {
                        IslandRadioRow(symbol: "wand.and.stars",
                                       title: "Auto",
                                       subtitle: "Let rules pick the right island at the right time.",
                                       selected: store.selectedKey == IslandDesignsStore.autoKey,
                                       last: store.entries.isEmpty,
                                       enabled: !store.isBusy) {
                            Task { await store.select(IslandDesignsStore.autoKey) }
                        }
                        ForEach(Array(store.entries.enumerated()), id: \.element.id) { index, entry in
                            IslandRadioRow(symbol: entry.iconName,
                                           title: entry.name,
                                           subtitle: entry.subtitle,
                                           selected: store.selectedKey == entry.id,
                                           last: index == store.entries.count - 1,
                                           enabled: !store.isBusy) {
                                Task { await store.select(entry.id) }
                            }
                        }
                    }
                    .padding(.bottom, 26)

                    GlassSectionLabel("In the Auto rotation")
                    GlassGroup {
                        ForEach(Array(store.entries.enumerated()), id: \.element.id) { index, entry in
                            IslandManageRow(entry: entry,
                                            busy: store.isBusy,
                                            last: index == store.entries.count - 1,
                                            onEnabled: { on in
                                                Task { await store.setEnabled(entry.id, on) }
                                            },
                                            onPriority: { value in
                                                Task { await store.setPriority(entry.id, value) }
                                            },
                                            onDelete: store.canDelete(entry)
                                                ? { deleting = entry } : nil)
                        }
                    }

                    if store.customEntries.isEmpty {
                        IslandEmptyHint().padding(.top, 18)
                    }
                }
                .padding(.horizontal, 20)
                .padding(.top, 8)
                .padding(.bottom, 32)
            }
            .refreshable { await store.refresh() }
        }
    }
}

/// One radio row: frosted glyph, title + subtitle, and the selection mark.
struct IslandRadioRow: View {
    let symbol: String
    let title: String
    let subtitle: String
    let selected: Bool
    let last: Bool
    let enabled: Bool
    let onSelect: () -> Void

    var body: some View {
        VStack(spacing: 0) {
            Button(action: onSelect) {
                HStack(spacing: 14) {
                    GlassCircleIcon(symbol: symbol)
                    VStack(alignment: .leading, spacing: 2) {
                        Text(title)
                            .font(.system(size: 15, weight: .semibold))
                            .foregroundStyle(JcTheme.text)
                        Text(subtitle)
                            .font(.system(size: 12))
                            .foregroundStyle(JcTheme.muted)
                            .lineLimit(2)
                            .multilineTextAlignment(.leading)
                    }
                    Spacer(minLength: 8)
                    Image(systemName: selected ? "largecircle.fill.circle" : "circle")
                        .font(.system(size: 20))
                        .foregroundStyle(selected ? JcTheme.primaryBlue : JcTheme.muted)
                }
                .padding(.horizontal, 14)
                .padding(.vertical, 12)
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .disabled(!enabled)
            if !last {
                Rectangle().fill(JcTheme.glassBorder).frame(height: 1).padding(.leading, 68)
            }
        }
    }
}

/// One management row: name, a priority nudger, delete (custom designs only) and
/// the Auto-rotation switch.
struct IslandManageRow: View {
    let entry: IslandCatalogEntry
    let busy: Bool
    let last: Bool
    let onEnabled: (Bool) -> Void
    let onPriority: (Int) -> Void
    let onDelete: (() -> Void)?

    var body: some View {
        VStack(spacing: 0) {
            HStack(spacing: 10) {
                Text(entry.name)
                    .font(.system(size: 15, weight: .semibold))
                    .foregroundStyle(JcTheme.text)
                    .lineLimit(1)
                Spacer(minLength: 4)
                priority
                if let onDelete {
                    Button(action: onDelete) {
                        Image(systemName: "trash")
                            .font(.system(size: 15))
                            .foregroundStyle(JcTheme.muted)
                            .frame(width: 34, height: 34)
                            .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)
                    .disabled(busy)
                    .accessibilityLabel("Delete \(entry.name)")
                }
                Toggle("", isOn: Binding(get: { entry.enabled }, set: onEnabled))
                    .labelsHidden()
                    .tint(JcTheme.primaryBlue)
                    .disabled(busy)
                    .accessibilityLabel("\(entry.name) in Auto rotation")
            }
            .padding(.horizontal, 14)
            .padding(.vertical, 8)
            if !last {
                Rectangle().fill(JcTheme.glassBorder).frame(height: 1).padding(.leading, 14)
            }
        }
    }

    /// Priority decides which enabled entry Auto picks when several match, so it
    /// needs to be adjustable — the Flutter screen only showed it in the subtitle.
    private var priority: some View {
        HStack(spacing: 2) {
            Button { onPriority(max(0, entry.priority - 1)) } label: {
                Image(systemName: "minus").font(.system(size: 11, weight: .bold))
                    .frame(width: 24, height: 24).contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .disabled(busy || entry.priority <= 0)
            Text("\(entry.priority)")
                .font(.system(size: 12, weight: .bold, design: .monospaced))
                .foregroundStyle(JcTheme.text)
                .frame(minWidth: 18)
            Button { onPriority(min(99, entry.priority + 1)) } label: {
                Image(systemName: "plus").font(.system(size: 11, weight: .bold))
                    .frame(width: 24, height: 24).contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .disabled(busy || entry.priority >= 99)
        }
        .foregroundStyle(JcTheme.muted)
        .background(JcTheme.glassFill, in: Capsule())
        .overlay(Capsule().strokeBorder(JcTheme.glassBorder, lineWidth: 1))
        .accessibilityLabel("\(entry.name) priority \(entry.priority)")
    }
}

/// The "ask Jarvis for one" nudge, shown when nothing custom exists yet.
struct IslandEmptyHint: View {
    var body: some View {
        Text("No custom designs yet. Ask Jarvis to make one — e.g. “make a Dynamic "
           + "Island that shows my next meeting”.")
            .font(.system(size: 13))
            .foregroundStyle(JcTheme.muted)
            .fixedSize(horizontal: false, vertical: true)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(16)
            .background(JcTheme.glassFill, in: RoundedRectangle(cornerRadius: 16, style: .continuous))
            .overlay(RoundedRectangle(cornerRadius: 16, style: .continuous)
                .strokeBorder(JcTheme.glassBorder, lineWidth: 1))
    }
}

/// A soft red banner for a catalog that wouldn't load.
struct IslandErrorCard: View {
    let message: String

    var body: some View {
        Text(message)
            .font(.system(size: 13))
            .foregroundStyle(JcTheme.text)
            .fixedSize(horizontal: false, vertical: true)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(14)
            .background(JcTheme.danger.opacity(0.12),
                        in: RoundedRectangle(cornerRadius: 14, style: .continuous))
            .overlay(RoundedRectangle(cornerRadius: 14, style: .continuous)
                .strokeBorder(JcTheme.danger.opacity(0.3), lineWidth: 1))
    }
}
